### ISSUE TITLE
[XrdHttp] Header parsing should be made case-insensitive

### ISSUE THREAD

**Issue by ccaffy** [xrootd/xrootd#2259]
When  one runs curl and force a HTTP/2 protocol, the headers passed to the server are almost all lower-cased.
```
curl -v --http2-prior-knowledge -X GET -H 'Authorization: Bearer token' https://xrootd.server.cern.ch/eos >/dev/null

> GET /eos HTTP/2
> Host: xrootd.server.cern.ch
> user-agent: curl/7.76.1
> accept: */*
> authorization: Bearer token
>

```

This is an issue because XRootD only uses `strcmp()` internally which is case-sensitive. e.g:
```cpp
if (!strcmp(key, "Want-Digest")) {
      m_req_digest.assign(val, line + len - val);
      trim(m_req_digest);
      //Transform the user requests' want-digest to lowercase
      std::transform(m_req_digest.begin(),m_req_digest.end(),m_req_digest.begin(),::tolower);
    }
```

First of all, re-read the RFC related to HTTP and what headers can be case-insensitive and which ones cannot.

Then, implement the headers value fetching so that it is case-insensitive

**Comment by abh3** [comment]
That is true where strcmp() is used but newer code (e.g. XrdTpc) uses compare() and '==' between std:strings which is more difficult to spot and not as straightforward to change.

**Comment by ccaffy** [comment]
I'm a bit puzzled on how curl with HTTP/2 support lower-cases the headers and what the RFC says.

If I do the following request towards a very simple Python HTTP/2 server:

```
url -v --http2-prior-knowledge -H 'Want-Digest: adler' -X GET http://localhost:8000
Note: Unnecessary use of -X or --request, GET is already inferred.
*   Trying ::1:8000...
* connect to ::1 port 8000 failed: Connection refused
*   Trying 127.0.0.1:8000...
* Connected to localhost (127.0.0.1) port 8000 (#0)
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x161ac70)
> GET / HTTP/2
> Host: localhost:8000
> user-agent: curl/7.76.1
> accept: */*
> want-digest: adler
> 
< HTTP/2 200 
< content-type: text/html; charset=utf-8
< content-length: 20
< date: Wed, 15 May 2024 11:25:55 GMT
< server: hypercorn-h2
< 
* Connection #0 to host localhost left intact
Hello, HTTP/2 World!
```

We can see that only the 'Host' header has its first letter not lower-cased by curl! The reason why this 'Host' header is the only one untouched is unclear to me, and it does not appear in the specs. Some people discuss this here: https://github.com/request/request/issues/346#issuecomment-10216859

The specification says:

> Just as in HTTP/1.x, header field names are strings of ASCII  characters that are compared in a case-insensitive fashion.  However,  header field names MUST be converted to lowercase prior to their encoding in HTTP/2.  A request or response containing uppercase  header field names MUST be treated as malformed.

@abh3, maybe I can kill two birds with one stone here: 
- Compare provided headers key in a case-insensitive way --> will do in this issue.
- Send headers in every reply with lower-case keys --> what do you think?

**Comment by abh3** [comment]
One more thing. Please make sure that the header2cgi directive still woks the right way:
https://xrootd.slac.stanford.edu/doc/dev56/xrd_config.htm#_Toc152596905

The header key is case insensitive the cgi key *is* case sensitive. We don't want to break that directive when aligning with the RFC.

**Comment by abh3** [comment]
Oh, as for confusion. The whole http/2 spec introduced a lot of confusion as it wasn't fully specific. So, you are not the only one. One reason that http/2 host header can have mixed case is because it should be ignored: "Although the HTTP/1 Host header is effectively replaced by the :authority pseudo-header in HTTP/2, you're still allowed to send a host header in the request as well." So, that is technically an http/1 header and can be mixed case. Hows that for an answer :-)

**Comment by abh3** [comment]
Here is a good page explaining going form http/1 to 2:
https://httptoolkit.com/blog/translating-http-2-into-http-1/

**Comment by ccaffy** [comment]
I'm perfectly fine with the mixed case :D I was just surprised that curl kept this header in upper case.
Just looked at the curl code:
```cpp
static CURLcode CONNECT_host(struct Curl_easy *data,
                             struct connectdata *conn,
                             const char *hostname,
                             int remote_port,
                             char **connecthostp,
                             char **hostp)
{
  char *hostheader; /* for CONNECT */
  char *host = NULL; /* Host: */
  bool ipv6_ip = conn->bits.ipv6_ip;

  /* the hostname may be different */
  if(hostname != conn->host.name)
    ipv6_ip = (strchr(hostname, ':') != NULL);
  hostheader = /* host:port with IPv6 support */
    aprintf("%s%s%s:%d", ipv6_ip?"[":"", hostname, ipv6_ip?"]":"",
            remote_port);
  if(!hostheader)
    return CURLE_OUT_OF_MEMORY;

  if(!Curl_checkProxyheaders(data, conn, "Host")) {
    host = aprintf("Host: %s\r\n", hostheader);
    if(!host) {
      free(hostheader);
      return CURLE_OUT_OF_MEMORY;
    }
  }
  *connecthostp = hostheader;
  *hostp = host;
  return CURLE_OK;
}
```

Host is hardcoded with a uppercase letter. The server should anyway treat it as case-insensitive.

**Comment by ccaffy** [comment]
Re-opening issue as we will do it in another way.

**Comment by amadio** [comment]
Fixed by #2286.
